inspector/src/components/graph/canvasGraphService.ts

```diff
@@ -416,9 +459,131 @@ export class CanvasGraphService {
      */
     private _removeEventListeners(canvas: HTMLCanvasElement) {
         canvas.removeEventListener("wheel", this._handleZoom);
+        canvas.removeEventListener("mousemove", this._handleDataHover);
         canvas.removeEventListener("mousedown", this._handlePanStart);
+        canvas.removeEventListener("mouseleave", this._handleStopHover);
         canvas.ownerDocument.removeEventListener("mouseup", this._handlePanStop);
     }
+    
+    /**
+     * Handles what to do when we are hovering over the canvas and not panning.
+     * 
+     * @param event A reference to the event to be handled. 
+     */
+    private _handleDataHover = (event: MouseEvent) => {
+        if (this._panPosition) {
+            // we don't want to do anything if we are in the middle of panning
+            return;
+        }
+
+        this._hoverPosition = event.clientX;
+
+        // then draw the tooltip.
+        this._drawTooltip(this._hoverPosition, this._drawableArea);
+    }
+
+    /**
+     * Handles what to do when we stop hovering over the canvas.
+     */
+    private _handleStopHover = () => {
+        this._hoverPosition = null;
+    }
+
+    /**
+     * Draws the tooltip given the area it is allowed to draw in and the current pixel position.
+     * 
+     * @param pixel the position of the mouse cursor in pixels. 
+     * @param drawableArea  the available area we can draw in.
+     */
+    private _drawTooltip(pixel: number | null, drawableArea: IGraphDrawableArea) {
+        const {_ctx : ctx} = this;
+
+        if (pixel === null || !ctx || !ctx.canvas) {            
+            return;
+        }
+
+        // first convert the mouse position in pixels to a timestamp.
+        const {left: start, right: end} = ctx.canvas.getBoundingClientRect();
+        const inferredTimestamp = this._getNumberFromPixel(pixel, this._globalTimeMinMax, start, end);
+        
+        const results: IPerfTooltip[] = [];
+
+        // get the closest timestamps to the target timestamp, and store the appropriate meta object.
+        this.datasets.forEach((dataset: IPerfDataset) => {
+            if (!!dataset.hidden || dataset.data.length === 0) {
+                return;
+            }
+
+            const closestIndex = this._getClosestPointToTimestamp(dataset, inferredTimestamp);
+            
+            results.push({
+                            text: `${dataset.id}: ${dataset.data[closestIndex].value.toFixed(2)}`,
+                            color: dataset.color ?? defaultColor,
+                        });
+        });
+
+        const longestText = results.reduce((longestSoFar: string, result) => {

```

Comment:
This goes over all the results again, why not doing it in the previous foreach ?

Reply:
Hmm I'll see if i can do that!

Reply:
Done!